[WPE] WPE Platform: add initial API tests for wayland platform

carlosgcampos opened the pull request:
#### 4141c4843eadde5c961cf70ac4e86cc3d9f986ae
<pre>
[WPE] WPE Platform: add initial API tests for wayland platform
<a href="https://bugs.webkit.org/show_bug.cgi?id=294671">https://bugs.webkit.org/show_bug.cgi?id=294671</a>

Reviewed by Carlos Alberto Lopez Perez.

Add an intial test to check we can create a WPEDisplayWayland and
connect to the wayland display. The script run-wpe-tests now detects if
a WPE Platform Wayland test is going to be run and it starts a weston
driver to ensure there&apos;s a wayland display to connect.

Canonical link: <a href="https://commits.webkit.org/296508@main">https://commits.webkit.org/296508@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b1254a45afed0970536494e867f18880591daa9f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/108758 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/28419 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/18843 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/113969 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/59142 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/110721 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/29108 "Build is in progress. Recent messages:OS: Sequoia (15.5), Xcode: 16.4; Skipping applying patch since patch_id isn't provided; Checked out pull request; Compiled WebKit (warnings)") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/36983 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/82630 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/111706 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/23127 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/97961 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/63067 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/22548 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/16098 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/58670 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/92497 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/16147 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/117090 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/35812 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/26443 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/91653 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/36185 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/94228 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/91459 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/36356 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/14110 "Passed tests") | [❌ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/31699 "Hash b1254a45 for PR 46900 does not build (failure)") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/17558 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/35712 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/35422 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/38769 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/37100 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

clopez commented on Tools/glib/api_test_runner.py:
```diff
@@ -178,12 +183,16 @@ def _use_wpe_legacy_api(self):
+        env = self._test_env
+
+        if self.is_wpe_platform_wayland_test(test_program):
+            env = self._weston_env
```
env = self._weston_env if self.is_wpe_platform_wayland_test(test_program) else self._test_env

clopez commented on Tools/glib/api_test_runner.py:
```diff
@@ -314,11 +326,24 @@ def run_tests(self):
+        # Skip WPE Platform Wayland tests if weston is not available.
+        if number_of_wpe_platform_wayland_tests > 0:
+            if not WestonDriver.check_driver(self._port):
+                self._tests = [test for test in self._tests if not self.is_wpe_platform_wayland_test(test)]
```
We should print a warning here like do when the Qt tests are skipped because of not having a working GPU

clopez commented on Tools/glib/api_test_runner.py:
```diff
@@ -314,11 +326,24 @@ def run_tests(self):
         # some tests might be skipped while setting up the test environment.
         self._tests = [test for test in self._tests if self._should_run_test_program(test)]
+        number_of_qt_tests = 0
+        number_of_wpe_platform_wayland_tests = 0
```
I suggest to change this variable name to `number_of_remaining_wpe_platform_wayland_tests` because it reflects better for what is used

clopez commented on Tools/glib/api_test_runner.py:
```diff
@@ -330,7 +355,17 @@ def run_tests(self):
+
+                if number_of_wpe_platform_wayland_tests == 0 and self._weston is not None:
+                    self._weston.stop()
+                    self._weston = None
```
this block should be moved into the `finally` (after the `try` (out of the for loop)) so it gets executed always, even when there is an exception so we ensure to always close weston and not leave left-over process.

Actually I think you don't need to count the number of remaining wayland tests. You can stop always the weston driver when is defined (is not none). That simplifies the logic and you can keep the original variable name `number_of_wpe_platform_wayland_tests`

Even better: just stop weston on `_tear_down_testing_environment()` which is what the finally block calls

```
    def _tear_down_testing_environment(self):
        if self._driver:
            self._driver.stop()
        if self._weston is not None:
            self._weston.stop()
            self._weston = None
```

carlosgcampos replied:
This patch already does that. But it also tries to stop weston early when we know that is no longer needed.

clopez replied:
allright, i missed that, sorry.

what is the reason to stop weston early?

carlosgcampos replied:
Just because we know it's no longer needed

clopez replied:
yeah, but it doesn't hurt either to wait until the runner ends to stop it, and it makes the code a bit more complex. Not sure if is worth such early optimization.

clopez replied:
anyway, is fine in any case.
r=me but please check the other comments before landing, specially the one about printing a warning when the tests can't run because weston is not available

carlosgcampos replied:
Nah, I agree it isn't probably worth it, so let's just keep weston running until teardown

carlosgcampos committed:
[WPE] WPE Platform: add initial API tests for wayland platform
https://bugs.webkit.org/show_bug.cgi?id=294671

Reviewed by Carlos Alberto Lopez Perez.

Add an intial test to check we can create a WPEDisplayWayland and
connect to the wayland display. The script run-wpe-tests now detects if
a WPE Platform Wayland test is going to be run and it starts a weston
driver to ensure there's a wayland display to connect.

Canonical link: https://commits.webkit.org/296508@main
